File: app/src/main/java/hibernate/v2/testyourandroid/ui/fragment/AppInfoPackageFragment.java
package hibernate.v2.testyourandroid.ui.fragment;

import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import hibernate.v2.testyourandroid.C;
import hibernate.v2.testyourandroid.R;
import hibernate.v2.testyourandroid.model.AppItem;
import hibernate.v2.testyourandroid.model.InfoItem;
import hibernate.v2.testyourandroid.ui.adapter.InfoItemAdapter;

public class AppInfoPackageFragment extends BaseFragment {
	private List<InfoItem> list = new ArrayList<>();
	private InfoItemAdapter adapter;

	@BindView(R.id.rvlist)
	RecyclerView recyclerView;
	private AppItem appItem;
	private PackageManager packageManager;
	private PackageInfo packageInfo;

	public AppInfoPackageFragment() {
		// Required empty public constructor
	}

	public static AppInfoPackageFragment newInstance(AppItem appItem) {
		AppInfoPackageFragment fragment = new AppInfoPackageFragment();
		Bundle args = new Bundle();
		args.putParcelable("APP", appItem);
		fragment.setArguments(args);
		return fragment;
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		// Inflate the layout for this fragment
		View rootView = inflater.inflate(R.layout.fragment_info_listview, container, false);
		ButterKnife.bind(this, rootView);
		return rootView;
	}

	@Override
	public void onViewCreated(View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		recyclerView.setLayoutManager(
				new LinearLayoutManager(mContext,
						LinearLayoutManager.VERTICAL, false)
		);
		init();
	}

	private void init() {
		Bundle bundle = getArguments();
		if (bundle != null) {
			appItem = bundle.getParcelable("APP");

			try {
				packageManager = mContext.getPackageManager();
				packageInfo = packageManager.getPackageInfo(appItem.getPackageName(), 0);

				list = new ArrayList<>();
				String[] stringArray = getResources().getStringArray(R.array.app_package_string_array);

				for (int i = 0; i < stringArray.length; i++) {
					list.add(new InfoItem(stringArray[i], getData(i)));
				}

				adapter = new InfoItemAdapter(list);

				recyclerView.setAdapter(adapter);
			} catch (PackageManager.NameNotFoundException e) {
				C.notAppFound(mContext);
			}
		} else {
			C.notAppFound(mContext);
		}
	}

	private String getData(int j) {
		try {
			Date date;
			DateFormat formatter = SimpleDateFormat.getDateTimeInstance();
			switch (j) {
				case 0:
					return (packageInfo.applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) != 0
							? "System Package" : "User-Installed Package";
				case 1:
					return packageInfo.packageName;
				case 2:
					return String.valueOf(packageInfo.versionCode);
				case 3:
					return packageInfo.versionName;
				case 4:
					date = new Date(packageInfo.firstInstallTime);
					return formatter.format(date);
				case 5:
					date = new Date(packageInfo.lastUpdateTime);
					return formatter.format(date);
				case 6:
					return packageInfo.applicationInfo.dataDir;
				case 7:
					return packageInfo.applicationInfo.publicSourceDir;
				case 8:
					if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
						return String.valueOf(packageInfo.applicationInfo.minSdkVersion);
					} else {
						return getString(R.string.notsupport_android_24);
					}
				case 9:
					return String.valueOf(packageInfo.applicationInfo.targetSdkVersion);
				default:
					return "N/A";
			}
		} catch (Exception e) {
			return "N/A";
		}
	}
}

File: app/src/main/java/hibernate/v2/testyourandroid/ui/appinfo/AppInfoPackageFragment.kt
package hibernate.v2.testyourandroid.ui.appinfo

import android.content.pm.PackageInfo
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.pm.PackageInfoCompat
import hibernate.v2.testyourandroid.R
import hibernate.v2.testyourandroid.databinding.FragmentInfoListviewBinding
import hibernate.v2.testyourandroid.model.AppItem
import hibernate.v2.testyourandroid.model.InfoItem
import hibernate.v2.testyourandroid.ui.appinfo.AppInfoFragment.Companion.ARG_APP
import hibernate.v2.testyourandroid.ui.base.BaseFragment
import hibernate.v2.testyourandroid.ui.base.InfoItemAdapter
import hibernate.v2.testyourandroid.util.Utils.notAppFound
import hibernate.v2.testyourandroid.util.ext.isSystemPackage
import java.text.SimpleDateFormat
import java.util.Date

class AppInfoPackageFragment : BaseFragment<FragmentInfoListviewBinding>() {

    override fun getViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): FragmentInfoListviewBinding =
        FragmentInfoListviewBinding.inflate(inflater, container, false)

    private lateinit var packageInfo: PackageInfo

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        arguments?.getParcelable<AppItem>(ARG_APP)?.let { appItem ->
            try {
                packageInfo = context?.packageManager?.getPackageInfo(appItem.packageName, 0)
                    ?: throw Exception()

                val stringArray = resources.getStringArray(R.array.app_package_string_array)
                val list = stringArray.mapIndexed { index, s -> InfoItem(s, getData(index)) }
                viewBinding!!.rvlist.adapter = InfoItemAdapter().apply {
                    setData(list)
                }
            } catch (e: Exception) {
                notAppFound(activity)
            }
        } ?: run {
            notAppFound(activity)
        }
    }

    private fun getData(j: Int): String {
        return try {
            val date: Date
            val formatter = SimpleDateFormat.getDateTimeInstance()
            when (j) {
                0 -> if (packageInfo.isSystemPackage()) "System Package" else "User-Installed Package"
                1 -> packageInfo.packageName
                2 -> PackageInfoCompat.getLongVersionCode(packageInfo).toString()
                3 -> packageInfo.versionName ?: throw Exception()
                4 -> {
                    date = Date(packageInfo.firstInstallTime)
                    formatter.format(date) ?: throw Exception()
                }

                5 -> {
                    date = Date(packageInfo.lastUpdateTime)
                    formatter.format(date) ?: throw Exception()
                }

                6 -> packageInfo.applicationInfo?.dataDir ?: throw Exception()
                7 -> packageInfo.applicationInfo?.publicSourceDir ?: throw Exception()
                8 -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    packageInfo.applicationInfo?.minSdkVersion?.toString() ?: throw Exception()
                } else {
                    getString(R.string.ui_not_support_android_version, "7.0")
                }

                9 -> packageInfo.applicationInfo?.targetSdkVersion?.toString() ?: throw Exception()
                else -> "N/A"
            }
        } catch (e: Exception) {
            "N/A"
        }
    }

    companion object {
        fun newInstance(appItem: AppItem?): AppInfoPackageFragment {
            val fragment = AppInfoPackageFragment()
            val args = Bundle()
            args.putParcelable(ARG_APP, appItem)
            fragment.arguments = args
            return fragment
        }
    }
}
